src/main/java/com/iota/iri/TransactionValidator.java

```diff
@@ -179,14 +179,18 @@ public boolean checkSolidity(Hash hash, boolean milestone, int maxProcessedTrans
                 final TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hashPointer);
                 if(!transaction.isSolid()) {
                     if (transaction.getType() == TransactionViewModel.PREFILLED_SLOT && !hashPointer.equals(Hash.NULL_HASH)) {
-                        transactionRequester.requestTransaction(hashPointer, milestone);
                         solid = false;
-                        break;
-                    } else {
-                        if (solid) {
-                            nonAnalyzedTransactions.offer(transaction.getTrunkTransactionHash());
-                            nonAnalyzedTransactions.offer(transaction.getBranchTransactionHash());
+
+                        if (milestone && !transactionRequester.containsMilestoneRequest(hashPointer)) {
+                            transactionRequester.requestTransaction(hashPointer, milestone);
+                            break;
+                        } else if (!milestone && !transactionRequester.contains(hashPointer)) {
```

Comment:
What's the point of these extra ifs if `requestTransaction` already does:
```
                if(milestone) {
                    transactionsToRequest.remove(hash);
                    milestoneTransactionsToRequest.add(hash);
```

Reply:
The point of these changes is not to "upgrade" a request to a milestone request (which is what the transactionRequester does in your mentioned code part), but to ONLY "BREAK" if we didnt request the missing transaction as part of the correct set yet.

So in case we call checkSolidity multiple times without having received the answer to our previous request yet, we will queue additional transaction requests for additional transactions that are ALSO missing. Before we always stopped at the same transaction requesting it again and again and again and only progressing to further transactions when we received an answer.

This little tweak is what made the sync spead go up 10 fold.

Reply:
:100: